providers/mastodon.go

```diff
@@ -0,0 +1,75 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"net/url"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// MastodonProvider represents a Mastodon based Identity Provider
+type MastodonProvider struct {
+	*ProviderData
+}
+
+var _ Provider = (*MastodonProvider)(nil)
+
+const (
+	mastodonProviderName = "Mastodon"
+	mastodonDefaultScope = "read:accounts"
+)
+
+// NewMastodonProvider initiates a new MastodonProvider
+func NewMastodonProvider(p *ProviderData, opts options.MastodonOptions) (*MastodonProvider, error) {
+	mastodonURL, err := url.Parse(opts.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	loginURL := mastodonURL.JoinPath("oauth/authorize")
+	redeemURL := mastodonURL.JoinPath("oauth/token")
+	verifyCredsURL := mastodonURL.JoinPath("api/v1/accounts/verify_credentials")
+
+	p.setProviderDefaults(providerDefaults{
+		name:        mastodonProviderName,
+		loginURL:    loginURL,
+		redeemURL:   redeemURL,
+		profileURL:  verifyCredsURL,
+		validateURL: verifyCredsURL,
+		scope:       mastodonDefaultScope,
+	})
+	p.getAuthorizationHeaderFunc = makeOIDCHeader
+
+	return &MastodonProvider{ProviderData: p}, nil
+}
+
+// GetEmailAddress returns the Mastodon account name.
+// For users local to this instance, this will not include a domain name.
+func (p *MastodonProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
```

Comment:
This provider impls `GetEmailAddress` because that's what the DO provider did, but I see `EnrichSession` is actually the new way of going about things here so I'm happy to redo this part.